ASQ: Scale Out scenario with secure connection strings

Scooletz opened the pull request:
Connects to https://github.com/Particular/PlatformDevelopment/issues/871
Addresses https://github.com/Particular/NServiceBus.AzureStorageQueues/issues/119

Scooletz commented:
Ping @SeanFeldman Could you review it?

SeanFeldman commented on nservicebus/azure-storage-queues/configuration.md:
```diff
@@ -135,13 +135,7 @@ This feature can be enabled by specifying `.UseAccountNamesInsteadOfConnectionSt
-
-NOTE: This feature is available in `NServiceBus.Azure.Transports.WindowsAzureStorageQueues` Versions 7 and above.
-
+For securing connection strings to storage accounts in Scale Out, see [Securing connection strings to storage accounts for Scale Out](/nservicebus/azure-storage-queues/multi-storageaccount-support.md#securing-connection-strings-to-storage-accounts-for-scale-out)
```
> in Scale Out

Did you mean "in Scaled Out mode"?

Scooletz replied:
Yes. Will update.

SeanFeldman commented on nservicebus/azure-storage-queues/multi-storageaccount-support.md:
```diff
@@ -59,6 +61,28 @@ Each endpoint uses its own Azure storage account, thereby increasing message thr
 
+## Securing connection strings to storage accounts for Scale Out
+
+[Securing connection strings to storage accounts](/nservicebus/azure-storage-queues/configuration.md#securing-connection-strings-to-storage-accounts) has been introduced in `NServiceBus.Azure.Transports.WindowsAzureStorageQueues` Version 7. When using a single account, securing connection string is limited to calling `.UseAccountNamesInsteadOfConnectionStrings()`. When Scaling Out is applied, sec ...
```
Here's a user perspective:

> UseAccountNamesInsteadOfConnectionStrings()

Reads as "account names should used instead of connections strings". Why should I distinguish between a single account or multiple ones? I want use names and period :) 

Following this description, when using multiple accounts one should not invoke `.UseAccountNamesInsteadOfConnectionStrings()`?

Scooletz replied:
I'm sorry but I'm not following. Are you saying that the method should be named:
`UseAccountsNamesInsteadOfConnectionStrings` 
or that there should be two methods with totally different names for single and multi account scenarios?

SeanFeldman replied:
> When using a single account, securing connection string is limited to calling `.UseAccountNamesInsteadOfConnectionStrings()`

Comparing this to [ASB](http://docs.particular.net/nservicebus/azure-service-bus/securing-connection-strings):

> To enable the same behavior for outgoing messages as well, use UseNamespaceNamesInsteadOfConnectionStrings() on the configuration API:

where "the same behavior" stands for:

> Versions 7 and above can map logical namespace name to a namespace connection string. If a valid mapping between a namespace name and a connection string doesn't exist a KeyNotFoundException is thrown.

This reads to me as two different outcomes of the similar feature. In ASB `UseAccountNamesInsteadOfConnectionStrings()` serves the purpose for outgoing messages. Where ASQ`s equivalent `UseAccountNamesInsteadOfConnectionStrings` is used to secure both incoming and outgoing messages for a single storage account case.

Thoughts?

SeanFeldman commented on nservicebus/azure-storage-queues/multi-storageaccount-support.md:
```diff
@@ -59,6 +61,28 @@ Each endpoint uses its own Azure storage account, thereby increasing message thr
+[Securing connection strings to storage accounts](/nservicebus/azure-storage-queues/configuration.md#securing-connection-strings-to-storage-accounts) has been introduced in `NServiceBus.Azure.Transports.WindowsAzureStorageQueues` Version 7. When using a single account, securing connection string is limited to calling `.UseAccountNamesInsteadOfConnectionStrings()`. When Scaling Out is applied, sec ...
+Consider the following example:
+
+- two endpoints `default` and `another` using different accounts for their input queues
```
`default` is an overloaded name :)
Would `endpoint-A` and `endpoint-B` serve the sample better?

Scooletz replied:
Will update the example and the naming.

SeanFeldman commented:
Ok, I think I know what confused me so much.
ASQ and ASB implement the same concept of secured connection strings, but API looks very different.

**ASQ**

```c#
var transport = endpointConfiguration.UseTransport<AzureStorageQueueTransport>();
transport.ConnectionString("connString") <== requires explicit name mapping
  .UseAccountNamesInsteadOfConnectionStrings(mapping =>
  {
    mapping.MapLocalAccount("default"); <== explicit name mapping
    mapping.MapAccount("another","anotherConnString");
  });
```

**ASB**

```c#
var transport = endpointConfiguration.UseTransport<AzureServiceBusTransport>();
transport.ConnectionString("connString"); // <== by convention is called `default` namespace name
transport.NamespacePartitioning()
  .AddNamespace("another", "anotherConnString");
```

As you can see above, ASB is a single step, where with ASQ its a 2 step process.

@Particular/azure-storage-queues-maintainers @Particular/azure-service-bus-maintainers should this part be a little ore consistent?

Scooletz commented:
The discussion about API, especially the mapping took place in this PR https://github.com/Particular/NServiceBus.AzureStorageQueues/pull/115#discussion_r62669964 I'm adding it for the sake of reference.
Frankly speaking I see no difference in number of steps. The only difference is that one uses lambda to add the mapping and to add the default name. We can't use `"default"` as the name would mean different accounts in the different endpoints and `ReplyTo` headers. Which `default` one would reply to? I see no way of using this convention for ASQ.

If we agree on changing this, I'm wondering if this change in API shouldn't be parked till the RC. 

@melkio Could you chime it?

Scooletz commented:
I'm parking my updates to this PR till we have settle a common goal for it ;-)

SeanFeldman commented:
> . We can't use "default" as the name would mean different accounts in the different endpoints and `ReplyTo` headers. 

Great point. Perhaps that's an issue with ASB then (@melkio)?

And perhaps we should have @Particular/nservicebus-maintainers involved in a broader discussion about `.ConnectionString()` supplied by [`TransportExtensions`](https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Unicast/Transport/Config/TransportExtensions.cs#L108) which doesn't account for transports to have a name (alias) associated with connection strings.

SimonCropp commented:
@Scooletz @SeanFeldman any update on this? need any help?

SeanFeldman commented:
@SimonCropp @Particular/azure-storage-queues-maintainers and @Particular/azure-service-bus-maintainers had a call and we've decided to first figure it out what it will look like for one transport and they apply the same for another. For a little bit this PR will be frozen.

indualagarsamy commented:
@Scooletz @SeanFeldman - Any updates or still status quo? What's the blocker?

SeanFeldman commented:
No development on this yet @indualagarsamy. @yvesgoeleven is going through the concept for ASB and once it's ready, we'll move forward.

indualagarsamy commented:
@SeanFeldman - any updates ?

SeanFeldman commented:
[ASB PR](https://github.com/Particular/NServiceBus.AzureServiceBus/pull/249) is in the workings. ASQ will follow it if PR is merged.

indualagarsamy commented:
@yvesgoeleven - I see that the PR that Sean mentioned has been merged. Any updates on this doco PR?

Scooletz commented:
Just wanted to say, that I've pushed the latest ASQ package and I'm staring to work on this.

Scooletz commented:
@Particular/azure-storage-queues-maintainers pls review. I've applied all the @SeanFeldman comments. This version is based on the newest ASQ that has its API aligned to ASB `AddressRouting`. This addresses and clarifies all the questions which were not for this PR per se. https://github.com/Particular/docs.particular.net/pull/1526#issuecomment-219075212

@Particular/docs-maintainers Pls review.

yvesgoeleven commented on nservicebus/azure-storage-queues/multi-storageaccount-support.md:
```diff
@@ -59,6 +61,28 @@ Each endpoint uses its own Azure storage account, thereby increasing message thr
 
 ![Scale out with multiple storage accounts](azure03.png "width=50%")
```
The transition from the above topic to the topic below reads very abruptly, I think the topic below needs a short introduction to make the transition, maybe start with a short introduction here what the problem exactly is (you only describe the solution, not the problem)

yvesgoeleven commented on nservicebus/azure-storage-queues/multi-storageaccount-support.md:
```diff
@@ -59,6 +61,28 @@ Each endpoint uses its own Azure storage account, thereby increasing message thr
+
+`DefaultAccountName(string name)` maps a name of an account used by this endpoint.
+`AccountRouting().AddAccount(string accountName, string connectionString)` adds a mapping for other accounts, enabling using `@` notation for destination addresses `queue_name@accountName`.
 
```
maybe add something here about the pitfall when using request/reply, that 'default account name' may imply something different to the replier than it does to the sender

yvesgoeleven commented:
👍 from me

Scooletz commented:
@Particular/docs-maintainers It's good from the @Particular/azure-storage-queues-maintainers point of view. Please review.

SimonCropp commented:
@Scooletz this looks good to go. only thing outstanding is @yvesgoeleven comment

Scooletz commented:
The last comment from @yvesgoeleven was https://github.com/Particular/docs.particular.net/pull/1526#issuecomment-231344320 and as far as I see I've addressed all the issues :) Could you provide a link to the comment you was referring to?

SimonCropp commented:
@Scooletz sorry seems to be fixed. Guessing a qwerk in the GH diff
